File: 贪吃蛇_重构/snake.cpp
#include "snake.h"

Snake::Snake() :pos({ {11,10},{10,10},{9,10} }), moveDir({ {0, -1}, { 0,1 }, { -1,0 }, { 1,0 } }), dir(Direction::right)
{

}

Snake & Snake::operator=(Snake & snake)
{
	this->pos = snake.pos;
	this->dir = snake.dir;
	return *this;
}

Snake::~Snake()
{

}

void Snake::move()
{
	//移动
	pos.pop_back(); //弹出尾巴
	std::pair<int, int> newHead(pos.front()); //用原来的头初始化新头
	//原来的头按照方向移动就是新头
	newHead.first += moveDir[dir].first;
	newHead.second += moveDir[dir].second;
	//把新头装上去成为真正的新头
	pos.push_front(newHead);
}

void Snake::display(bool isDisplay)
{
	Console *console = Console::getInstance();
	//遍历蛇身坐标然后挨个在相应的位置输出相应的字符
	for (auto it = pos.cbegin(); it != pos.cend(); it++)
	{
		try
		{
			console->moveCursor(it->first, it->second);
		}
		catch (std::exception) //超出控制台大小范围的不画蛇身
		{
			continue;
		}
		if (isDisplay)
		{
			printf("#");
		}
		else
		{
			printf(" ");
		}
	}
}

void Snake::changeDir()
{
	//转换方向
	if (_kbhit()) //如果有按键按下的话
	{
		switch (_getch()) //则获取按键并且判断将要转换的方向
		{
		case 'w':
		{
			if (dir != Direction::down)
			{
				dir = Direction::up;
			}
			break;
		}
		case 's':
		{
			if (dir != Direction::up)
			{
				dir = Direction::down;
			}
			break;
		}
		case 'a':
		{
			if (dir != Direction::right)
			{
				dir = Direction::left;
			}
			break;
		}
		case 'd':
		{
			if (dir != Direction::left)
			{
				dir = Direction::right;
			}
			break;
		}
		}
	}
}

bool Snake::isDead()
{
	//获取蛇头
	std::pair<int, int> head = pos.front();
	//判断蛇头是否撞到边界
	if (head.first < 0 || head.first >= Console::getInstance()->getWindowWidth() || head.second < 0 || head.second >= Console::getInstance()->getWindowHeight())
	{
		return true;
	}
	//判断蛇头是否撞到自己的身体
	else
	{
		//从蛇头之后第一个开始判断起（如果有的话）
		auto it = pos.cbegin();
		it++;
		for (; it != pos.cend(); it++)
		{
			//如果和蛇头重叠
			if (it->first == head.first&&it->second == head.second)
			{
				return true;
			}
		}
	}
	//否则没死
	return false;
}

bool Snake::isEatFood(Food food)
{
	//如果吃到了
	if (pos.front().first == food.getX() && pos.front().second == food.getY())
	{
		//获取尾巴
		std::pair<int, int> last = pos.back();
		//获取倒数第二节
		pos.pop_back();
		std::pair<int, int> lastSecond = pos.back();
		//重新把尾巴安上去
		pos.push_back(last);
		//新的尾巴
		std::pair<int, int> newLast{};
		//计算新的尾巴的位置
		newLast.first = last.first + last.first - lastSecond.first;
		newLast.second = last.second + last.second - lastSecond.second;
		//安装
		pos.push_back(newLast);
		return true;
	}
	else
	{
		return false;
	}
}

File: snake_go/snake/snake.go
package snake

import (
	"snake/console"
)

type Direction int

// 方向
const (
	Up Direction = iota
	Down
	Left
	Right
)

// 移动
var moveMap map[Direction]SnakePos = map[Direction]SnakePos{
	Up:    {R: -1, C: 0},
	Down:  {R: 1, C: 0},
	Left:  {R: 0, C: -1},
	Right: {R: 0, C: 1},
}

// 坐标
type SnakePos struct {
	R, C int // 行 列
}

func (pos *SnakePos) Add(pos2 *SnakePos) {
	pos.R += pos2.R
	pos.C += pos2.C
}
func (pos *SnakePos) Sub(pos2 *SnakePos) {
	pos.R -= pos2.R
	pos.C -= pos2.C
}

// 蛇
type Snake struct {
	Dir Direction  // 方向
	Pos []SnakePos // 蛇身坐标
}

func New(dir Direction, R, C, length int) *Snake {
	snake := &Snake{Dir: dir}
	snake.Pos = make([]SnakePos, length)
	switch dir {
	case Up:
		for i := range length {
			snake.Pos[i] = SnakePos{R + i, C}
		}
	case Down:
		for i := range length {
			snake.Pos[i] = SnakePos{R - i, C}
		}
	case Left:
		for i := range length {
			snake.Pos[i] = SnakePos{R, C + i}
		}
	default:
		for i := range length {
			snake.Pos[i] = SnakePos{R, C - i}
		}
	}
	return snake
}

func (snake *Snake) Move() {
	for i := len(snake.Pos) - 1; i > 0; i-- {
		snake.Pos[i] = snake.Pos[i-1]
	}
	v := moveMap[snake.Dir]
	snake.Pos[0].Add(&v)
}

func (snake *Snake) Draw(b byte) {
	for _, pos := range snake.Pos {
		console.MoveCursor(pos.R, pos.C)
		console.Draw(b)
	}
}

func (snake *Snake) Grow() {
	if len(snake.Pos) == 1 {
		pos := SnakePos{snake.Pos[0].R, snake.Pos[0].C}
		v := moveMap[snake.Dir]
		pos.Sub(&v)
		snake.Pos = append(snake.Pos, pos)
	} else {
		last := snake.Pos[len(snake.Pos)-1]
		pre := snake.Pos[len(snake.Pos)-2]
		v := SnakePos{last.R*2 - pre.R, last.C*2 - pre.C}
		snake.Pos = append(snake.Pos, v)
	}
}

// 蛇是否撞上自己
func (s *Snake) IsKnockSelf() bool {
	for i := 1; i < len(s.Pos); i++ {
		if s.Pos[0] == s.Pos[i] {
			return true
		}
	}
	return false
}
